storage/index/types.go

```diff
@@ -183,42 +179,306 @@ type BlockTickResult struct {
+// with metadata in a separate slice, this allows the documents slice to be
+// passed to the segment to batch insert without having to copy into a buffer
+// again.
+type WriteBatch struct {
+	opts   WriteBatchOptions
+	sortBy writeBatchSortBy
 
-func (w WriteBatchEntryByBlockStart) Len() int           { return len(w) }
-func (w WriteBatchEntryByBlockStart) Swap(i, j int)      { w[i], w[j] = w[j], w[i] }
-func (w WriteBatchEntryByBlockStart) Less(i, j int) bool { return w[i].BlockStart < w[j].BlockStart }
+	entries []WriteBatchEntry
+	docs    []doc.Document
+}
 
-// ForEachBlockStartFn is lambda to iterate over WriteBatchEntry(s) a single blockStart at a time.
-type ForEachBlockStartFn func(blockStart xtime.UnixNano, writes []WriteBatchEntry)
+type writeBatchSortBy uint
+
+const (
+	writeBatchSortByUnmarkedAndBlockStart writeBatchSortBy = iota
+	writeBatchSortByEnqueued
+)
+
+// WriteBatchOptions is a set of options required for a write batch.
+type WriteBatchOptions struct {
+	InitialCapacity int
+	IndexBlockSize  time.Duration
+}
+
+// NewWriteBatch creates a new write batch.
+func NewWriteBatch(opts WriteBatchOptions) *WriteBatch {
+	return &WriteBatch{
+		opts:    opts,
+		entries: make([]WriteBatchEntry, 0, opts.InitialCapacity),
+		docs:    make([]doc.Document, 0, opts.InitialCapacity),
+	}
+}
+
+// Append appends an entry with accompanying document.
+func (b *WriteBatch) Append(
+	entry WriteBatchEntry,
+	doc doc.Document,
+) {
+	// Set private WriteBatchEntry fields
+	entry.enqueuedIdx = len(b.entries)
+	entry.result = WriteBatchEntryResult{}
+
+	// Append
+	b.entries = append(b.entries, entry)
+	b.docs = append(b.docs, doc)
+}
+
+// ForEachWriteBatchEntryFn allows a caller to perform an operation for each
+// batch entry.
+type ForEachWriteBatchEntryFn func(
+	idx int,
+	entry WriteBatchEntry,
+	doc doc.Document,
+	result WriteBatchEntryResult,
+)
+
+// ForEach allows a caller to perform an operation for each batch entry.
+func (b *WriteBatch) ForEach(fn ForEachWriteBatchEntryFn) {
+	for idx, entry := range b.entries {
+		fn(idx, entry, b.docs[idx], entry.Result())
+	}
+}
+
+// ForEachWriteBatchByBlockStartFn allows a caller to perform an operation with
+// reference to a restricted set of the write batch for each unique block
+// start.
+type ForEachWriteBatchByBlockStartFn func(
+	blockStart time.Time,
+	batch *WriteBatch,
+)
+
+// ForEachUnmarkedBatchByBlockStart allows a caller to perform an operation
+// with reference to a restricted set of the write batch for each unique block
+// start for entries that have not been marked completed yet.
+func (b *WriteBatch) ForEachUnmarkedBatchByBlockStart(
+	fn ForEachWriteBatchByBlockStartFn,
+) {
+	// Ensure sorted correctly first
+	b.SortByUnmarkedAndIndexBlockStart()
+
+	// What we do is a little funky but least alloc intensive, essentially we mutate
+	// this batch and then restore the pointers to the original docs after.
```

Comment:
nit: could you add a comment to ForEachWriteBatchByBlockStartFn  indicating that this will break if `fn` does async operations on the batch

Reply:
Good call, will do.